Make version part of metadata.json

CasperWA opened the pull request:
#### Version in `metadata.json`
To be able to do versioning of AiiDA Lab apps properly, a `version` key is added to `metadata.json`.
The default is still "0.1-alpha".

~~In the new AiiDA Lab app's `__init__.py` file, the `__version__` is set from the new `version` value in `metadata.json`.
This is done in order to only ever have to change the version in a single place (`metadata.json`).~~
_Edit_: `__version__` is removed from `__init__.py`.

#### Minor fixes
Update pre-commit key (`sha` is now `rev`) and up versions of hooks.
Update `.gitignore`.
~~Update `.travis.yml` files with `--no-use-pep517` when doing `pip install`.~~
_Edit_: A larger rewrite of `.travis.yml` was undertaken in PR #7.

CasperWA committed:
Version is now part of `metadata.json`.

Update pre-commit key (`sha` is now `rev`) and hook versions.

Added extra helpful info text in post-cookiecutter hook.

Remove `__version__` from `__init__.py` and only have versioning in `metadata.json`.

CasperWA commented:
@ltalirz @yakutovicha I think I'll merge this at this point, unless you see an imminent reason not to do so?

ltalirz commented on .travis.yml:
```diff
@@ -6,7 +6,7 @@ before_install:
 
 install:
-- pip install cookiecutter
+- pip install --no-use-pep517 cookiecutter
```
is this needed?

CasperWA replied:
I guess not - just so used to it for aiida, hehe - will remove it, thanks.

ltalirz commented on {{cookiecutter.repo_name}}/__init__.py:
```diff
@@ -1,7 +1,20 @@
+
+path_to_metadata = path.join(path.dirpath(__file__), "metadata.json")
+with open(path_to_metadata, 'r') as fp:
+    metadata = load(fp)
```
hm.... so if someone deletes the metadata.json, the import of the package should fail?
We could be doing this in plugins and aiida-core as well - I guess one of the reasons we don't (and rather put a pre-commit hook) is because this seems rather fragile.
I would vote for the pre-commit hook.

CasperWA replied:
It is quite vulnerable, for sure.
In the context of AiiDA Lab, however, we would _always_ need a `metadata.json` though, right?

CasperWA replied:
Anyway, what would the pre-commit hook look like in this context? Do you have an example from AiiDA I can go to or similar?

ltalirz replied:
Sure, see e.g. here https://github.com/aiidateam/aiida-diff/blob/master/.pre-commit-config.yaml#L32

CasperWA replied:
What about doing both things?
My idea was to minimize the places it needs to be changed from two to one (metadata.json).

yakutovicha commented:
Ah, I am sorry, I haven't seen this PR at all. Can I have some time to review it as well?

CasperWA commented:
> Ah, I am sorry, I haven't seen this PR at all. Can I have some time to review it as well?

Sure, if you wish. I believe one review should be enough - but more eyes are better than fewer here. Usually :)

CasperWA commented:
New commit:

Added version check via pre-commit hook.
It will check the `version` key in `metadata.json` against `__version__` in `__init__.py`.

Moved `__init__.py` into subfolder to make the new version check work,
and also to separate the module from the general repo.

Added extra helpful info text in post-cookiecutter hook.

ltalirz commented:
@CasperWA moving the  `__init__.py` out of the top-level folder won't work: the app folder is imported by the app manager.

You are right, of course, that this is not the standard python packaging way. However, at this point, an AiiDA lab app *is not* a python package. When it does contain python packages (like the [widgets](https://github.com/aiidalab/aiidalab-widgets-base)), they can be packaged separately with a `setup.py` etc.

Thoughts on best practises for packaging apps with jupyter notebooks are welcome! It's not an easy problem, though - the colleagues over at mybinder.org are also trying to figure this out (see e.g. [here](https://mybinder.readthedocs.io/en/latest/config_files.html#config-files) for their current set of config files).

For now, perhaps the easiest solution would be to simply remove the `version` from the `__init__.py` (and get rid of the pre-commit hooks etc.)
@yakutovicha Do you know whether the version from the __init__.py is used anywhere?

Sorry, I should have thought about this earlier.

CasperWA commented:
> Thoughts on best practises for packaging apps with jupyter notebooks are welcome! It's not an easy problem, though - the colleagues over at mybinder.org are also trying to figure this out (see e.g. [here](https://mybinder.readthedocs.io/en/latest/config_files.html#config-files) for their current set of config files).

Right. So we should make an issue that deals with this question?

> For now, perhaps the easiest solution would be to simply remove the `version` from the `__init__.py` (and get rid of the pre-commit hooks etc.)

I think you're right. But that makes it crucial to have the `metadata.json` and `version` key in it, to allow for versioning logic when updating it through the AiiDA Lab App Store.

> @yakutovicha Do you know whether the version from the `__init__.py` is used anywhere?

Will push a new commit as soon as I hear back on this.

> Sorry, I should have thought about this earlier.

No worries :)

yakutovicha commented:
> @yakutovicha Do you know whether the version from the init.py is used anywhere?

not that I know

yakutovicha commented:
Another comment: Home App uses git versioning. It does not require any additional source of information

CasperWA commented:
> Another comment: Home App uses git versioning. It does not require any additional source of information

Ah - so the versioning in `metadata.json` is only for human-readable changes in the end. Well. Maybe not so important then?

We could have a hook that updates `metadata.json` with `__version__` as well, if they differ? This way one would only change it there. The danger being, if people don't install the hooks, then it will not update `metadata.json` and they may be confused to see that the version doesn't change in the registry/app store.

ltalirz commented on hooks/post_gen_project.py:
```diff
@@ -1,6 +1,22 @@
+
 print("""
-### IMPORTANT###
+### IMPORTANT ###
```
Proposal:

```
### NEXT STEPS ###

Enable version control:

    cd {{cookiecutter.repo_name}}
    git init

Enable code sanity checks on every commit:

    pip install pre-commit    
    pre-commit install

Register your app NOW by making a pull request to the AiiDA Lab App Registry
on https://github.com/aiidalab/aiidalab-registry
See this link also for a full list of keys for your metadata.json file.
```

CasperWA replied:
Done.
However, I have modified it a bit further and now reads:

```
Congratulations!
You have now taken the first step in creating your new AiiDA Lab App:

    {{cookiecutter.app_title}}

### NEXT STEPS ###

1. Open your new AiiDA Lab App folder

    cd {{cookiecutter.repo_name}}

2. Enable version control

    git init

3. Enable code sanity checks on every commit (recommended)

    pip install -U pre-commit
    pre-commit install

4. Register your app NOW by making a pull request to the AiiDA Lab App Registry
on https://github.com/aiidalab/aiidalab-registry
See this link also for a full list of keys for your metadata.json file.
```

ltalirz reviewed (changes requested):
Looks good!
Just one last proposal

ltalirz reviewed (approved):
approoooooooved!

CasperWA commented:
> approoooooooved!

w0000000t
